app/src/main/java/nz/org/cacophony/birdmonitor/TestRecordFragment.java

```diff
@@ -177,4 +192,124 @@ public void onReceive(Context context, Intent intent) {
+//        https://developer.android.com/training/permissions/requesting.html
+
+        boolean allPermissionsAlreadyGranted = true;
+
+        if (ContextCompat.checkSelfPermission(context,
+                Manifest.permission.WRITE_EXTERNAL_STORAGE)
+                != PackageManager.PERMISSION_GRANTED) {
+
+            allPermissionsAlreadyGranted = false;
+
+            //https://stackoverflow.com/questions/35989288/onrequestpermissionsresult-not-being-called-in-fragment-if-defined-in-both-fragm
+
+            requestPermissions(new String[]{
+                    Manifest.permission.WRITE_EXTERNAL_STORAGE}, PERMISSION_WRITE_EXTERNAL_STORAGE);
+
+        }
+
+        if (ContextCompat.checkSelfPermission(context,
+                Manifest.permission.RECORD_AUDIO)
+                != PackageManager.PERMISSION_GRANTED) {
+
+            allPermissionsAlreadyGranted = false;
+
+            requestPermissions(new String[]{
+                    Manifest.permission.RECORD_AUDIO}, PERMISSION_RECORD_AUDIO);
+
+        }
+
+        if (ContextCompat.checkSelfPermission(context,
+                Manifest.permission.ACCESS_FINE_LOCATION)
+                != PackageManager.PERMISSION_GRANTED) {
+
+            allPermissionsAlreadyGranted = false;
+
+            requestPermissions(new String[]{
+                    Manifest.permission.ACCESS_FINE_LOCATION}, PERMISSION_LOCATION);
+
+        }
+
+        return allPermissionsAlreadyGranted;
+    }
+
+    @Override
+    public void onRequestPermissionsResult(int requestCode, String[] permissions,
+                                           int[] grantResults) {
+        // BEGIN_INCLUDE(onRequestPermissionsResult)
+        if (requestCode == PERMISSION_WRITE_EXTERNAL_STORAGE) {
+            // Request for camera permission.
+            if (grantResults.length == 1 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
+                // Permission has been granted. Start recording
+                tvMessages.setText("WRITE_EXTERNAL_STORAGE permission granted");
+            } else {
+                tvMessages.setText("Do not have WRITE_EXTERNAL_STORAGE permission, You can NOT save recordings");
+            }
+        }
+
+        if (requestCode == PERMISSION_RECORD_AUDIO) {
+            // Request for camera permission.
+            if (grantResults.length == 1 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
+                // Permission has been granted. Start recording
+                tvMessages.setText("RECORD_AUDIO permission granted");
+            } else {
+                tvMessages.setText("Do not have RECORD_AUDIO permission, You can NOT record");
+            }
+        }
+
+        // May as well check GPS Location permission again
+
+        if (requestCode == PERMISSION_LOCATION) {
+            // Request for camera permission.
+            if (grantResults.length == 1 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
```

Comment:
extract this to a `wasGrantedPermission` function given that it's repeated 3 times here and in other files as well.

Reply:
Do you mean just this line:
  if (grantResults.length == 1 && grantResults[0] == PackageManager.PERMISSION_GRANTED) 

I can replace with

if (wasGrantedPermission(grantResults.length, grantResults[0])) {

 private boolean wasGrantedPermission(int grantResultsLength , int grantResult){
        if (grantResultsLength == 1 && grantResult == PackageManager.PERMISSION_GRANTED){
            return true;
        }else {
            return false;
        }
    }

Reply:
Something like that, except I would expect the `wasGrantedPermission` would take just `grantResults`, not the length and first item separately.